### PR TITLE
feat(gitlab): support getDigest for gitlab repositories

### PR DESCRIPTION
**Pull request by bastianccm**
## Changes:

Support `getDigest` for gitlab repositories.

## Context:

`gitlab-tags` currently does not support `getDigest` (github and bitbucket do). This change adds `getDigest` to `gitlab-tags`

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/gitlab-tags/index.ts [hunk 1]
```diff
@@ -69,3 +68,64 @@ export async function getReleases({
...
+  if (newValue) {
+    const url = joinUrlParts(
+      depHost,
+      `api/v4/projects`,
+      urlEncodedRepo,
+      `repository/commits/`,
+      newValue
+    );
+    const gitlabCommits = await gitlabApi.getJson<GitlabCommit>(url, {
+      paginate: true,
```
Comment (viceice) [inline lib/datasource/gitlab-tags/index.ts:104]: Are you sure about paging? i think we only need the first page for the id?

File: lib/datasource/gitlab-tags/index.ts [hunk 2]
```diff
@@ -69,3 +68,64 @@ export async function getReleases({
...
+    cacheNamespace,
+    getCacheKey(depHost, repo, 'commit')
+  );
+  // istanbul ignore if
+  if (cachedResult) {
+    return cachedResult;
+  }
+
+  const urlEncodedRepo = encodeURIComponent(repo);
+  let digest;
```
Comment (viceice) [inline lib/datasource/gitlab-tags/index.ts:93]:
```suggestion
  let digest: string;
```

File: lib/datasource/gitlab-tags/index.ts [hunk 3]
```diff
@@ -69,3 +68,64 @@ export async function getReleases({
...
+    digest = gitlabCommits.body.id;
+  } else {
+    const url = joinUrlParts(
+      depHost,
+      `api/v4/projects`,
+      urlEncodedRepo,
+      `repository/commits?per_page=1`
+    );
+    const gitlabCommits = await gitlabApi.getJson<GitlabCommit[]>(url, {
+      paginate: true,
```
Comment (viceice) [inline lib/datasource/gitlab-tags/index.ts:115]: I think we only need the first page? You're now fetching all commits one-by-one.

File: lib/datasource/go/digest.ts [hunk 4]
```diff
@@ -32,6 +33,9 @@ export async function getDigest(
     case bitbucket.id: {
       return bitbucket.getDigest(source, tag);
     }
+    case gitlab.id: {
+      return gitlab.getDigest(source, tag);
+    }
```
Comment (viceice) [inline lib/datasource/go/digest.ts:38]: i think this needs a new test?

Comment (bastianccm) [reply lib/datasource/go/digest.ts:38]: Yes, and actually the current spec for `getDigest` is not failing as expected:
https://github.com/renovatebot/renovate/blob/main/lib/datasource/go/digest.spec.ts#L52-L63

I'll fix and add a correct test

File: lib/datasource/go/digest.spec.ts [hunk 5]
```diff
@@ -49,16 +49,23 @@ describe('datasource/go/digest', () => {
...
-        { lookupName: 'gitlab.com/golang/text' },
+        { lookupName: 'gitlab.com/group/subgroup' },
         null
       );
-      expect(res).toBeNull();
+      expect(res).toMatchSnapshot();
+      expect(res).not.toBeNull();
+      expect(res).toBeDefined();
+      expect(res).toBe('abcdefabcdefabcdefabcdef');
       expect(httpMock.getTrace()).toMatchSnapshot();
```
Comment (viceice) [inline lib/datasource/go/digest.spec.ts:69]:
```suggestion
```
Obsolete 😉

File: lib/datasource/go/digest.spec.ts [hunk 6]
```diff
@@ -49,16 +49,23 @@ describe('datasource/go/digest', () => {
...
+        .scope('https://gitlab.com/')
+        .get('/api/v4/projects/group%2Fsubgroup/repository/commits?per_page=1')
+        .reply(200, [{ id: 'abcdefabcdefabcdefabcdef' }]);
       const res = await getDigest(
-        { lookupName: 'gitlab.com/golang/text' },
+        { lookupName: 'gitlab.com/group/subgroup' },
         null
       );
-      expect(res).toBeNull();
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/datasource/go/digest.spec.ts:65]:
```suggestion
```

The `toBe` will be enough, so no snapshot terst required

Comment (bastianccm) [reply lib/datasource/go/digest.spec.ts:65]: What about
```
expect(res).not.toBeNull();
expect(res).toBeDefined();
```
Ar they needed? `expect(res).toBe('abcdefabcdefabcdefabcdef');` should actually be sufficient for this test I think? (I've seen multiple expectations in other places so I did that too...)

Comment (bastianccm) [reply lib/datasource/go/digest.spec.ts]: I removed these expectations too

**Review by viceice — DISMISSED** [review]

**Comment by bastianccm** [comment]
@viceice could you make the linter run again? I hope I got the testcoverage right this time.

**Review by viceice — APPROVED** [review]

**Pull request merged**